pkg/generator/specification.go

```diff
@@ -113,56 +125,78 @@ func (s *SpecificationGenerator) generateIATPerGranularity(numberOfInvocations i
 				break
 			}
 		}
-		beginningIAT := sum - split
-		endIAT := iatResult[i] - beginningIAT
-		finalIAT := append([]float64{beginningIAT}, iatResult[i+1:]...)
-		finalIAT = append(finalIAT, iatResult[:i]...)
-		iatResult = append(finalIAT, endIAT)
-	} else {
-		iatResult = append([]float64{0.0}, iatResult...)
+
+		if i < len(iatResult) && i+1 < len(iatResult) {
+			beginningIAT := sum - split
+			endIAT := iatResult[i] - beginningIAT
+			finalIAT := append([]float64{beginningIAT}, iatResult[i+1:]...)
+			finalIAT = append(finalIAT, iatResult[:i]...)
+			iatResult = append(finalIAT, endIAT)
+		}
 	}
 
 	return iatResult, totalDuration
 }
 
+func getBlankTimeUnit(granularity common.TraceGranularity) float64 {
+	if granularity == common.MinuteGranularity {
+		return 60_000_000
+	} else {
+		return 1_000_000
+	}
+}
+
 // GenerateIAT generates IAT according to the given distribution. Number of minutes is the length of invocationsPerMinute array
-func (s *SpecificationGenerator) generateIAT(invocationsPerMinute []int, iatDistribution common.IatDistribution, shiftIAT bool, granularity common.TraceGranularity) (common.IATMatrix, common.ProbabilisticDuration) {
-	var IAT [][]float64
+func (s *SpecificationGenerator) generateIAT(invocationsPerMinute []int, iatDistribution common.IatDistribution,
+	shiftIAT bool, granularity common.TraceGranularity) (common.IATArray, []int, common.ProbabilisticDuration) {
+
+	var IAT []float64
+	var perMinuteCount []int
 	var nonScaledDuration []float64
 
+	accumulatedIdle := 0.0
+
 	numberOfMinutes := len(invocationsPerMinute)
 	for i := 0; i < numberOfMinutes; i++ {
-		minuteIAT, duration := s.generateIATPerGranularity(invocationsPerMinute[i], iatDistribution, shiftIAT, granularity)
+		minuteIAT, duration := s.generateIATPerGranularity(i, invocationsPerMinute[i], iatDistribution, shiftIAT, granularity)
+		if len(minuteIAT) == 0 {
+			accumulatedIdle += getBlankTimeUnit(granularity)
+			continue
+		} else if accumulatedIdle != 0 {
+			IAT = append(IAT, accumulatedIdle)
+			IAT = append(IAT, minuteIAT[1:]...)
+			perMinuteCount = append(perMinuteCount, len(minuteIAT)-1)
+			accumulatedIdle = 0.0
+		} else {
+			IAT = append(IAT, minuteIAT...)
+			perMinuteCount = append(perMinuteCount, len(minuteIAT))
+		}
```

Comment:
I think we can do much simpler with this (and restore the `generateIATPerGranularity` to the old version):
```suggestion
		minuteIAT, duration := s.generateIATPerGranularity(invocationsPerMinute[i], iatDistribution, shiftIAT, granularity)
		IAT[len(IAT)-1] += minuteIAT[0]
		IAT = append(IAT, minuteIAT[1:]...)
```

If we add `return []float64{getBlankTimeUnit}, 0.0` in case of no invocation in the period and start with `IAT:=[]float64{0.0}`.

On the minute border, real IAT is the last element of the previous minute + the first element of the next one. That's because the last element in the array is the interval between the last invocation and the end of the minute, while the first one is between the start of the minute and the first invocation. So, the method would report the interval between the arrivals. With the empty minutes, the last element of the IAT array would play the role of `accumulateIdle`, summing all the empty minutes between invocations.